README.md
# KalRTPC
 Kalman Filter program for RTPC12

##How to get these source code: 

 git clone git@github.com:jixie/KalRTPC.git 
 or
 git clone https://github.com:jixie/KalRTPC.git 

 This program depends on 3 other software packages:
 1) root, you can download it from "root.cern.ch", any version should be fine.
 2) clhep, I used 2.1.0.1 compiled with vc10 in windows xp, and 2.1.3.1 in CentOS_6.5
 3) KalmanFilter libraries originally from KEK, modified by me. 
 You can also get it from github:  
 git clone git@github.com:jixie/KalmanFilter.git 
 or
 git clone https://github.com:jixie/KalmanFilter.git 
 
 You need to have these 2 enviroment variables set: ${KALMANROOT} and ${CLHEP_BASE_DIR}

##How to compile?

 First of all, you need to define $rtpcsoft.  
 You can simply define it this way:  
   setenv rtpcsoft $home/rtpcsoft
 
 1) install root and clhep    
                      
  setenv ROOTSYS ${rtpcsoft}/root-5.28.00-x86_64                              
  setenv ROOTLIB $ROOTSYS/lib                                                
  setenv PATH ${ROOTSYS}/bin:${PATH}                                         
  setenv LD_LIBRARY_PATH ${ROOTSYS}/lib:${LD_LIBRARY_PATH}     

  setenv  CLHEP_BASE_DIR  ${rtpcsoft}/clhep-2.1.3.1-x86_64
  setenv  LD_LIBRARY_PATH ${CLHEP_BASE_DIR}/lib:${LD_LIBRARY_PATH}

 2) install KalmanFilter

  setenv KALMANROOT   ${rtpcsoft}/KalmanFilter  
  setenv LD_LIBRARY_PATH  ${KALMANROOT}/lib:$LD_LIBRARY_PATH

 cd KalmanFilter, change 'setup.csh' to define where you want to 
 install KalmanFilter package, then change to  KalmanFilter/src,
 type 'make clean; make;make install' to install it.

 3) compile kalRTPC
 
  cd kalRTPC;
  make clean;
  make



========================================================================
##To run this in windows:

vc10 project files are also available. I can provide g77 compiler and 
VC10 project files for clhep. You need to download and install vc10 version 
of root_v5.28 from "root.cern.ch".  Place KalmanFilter and KalRTPC into 
the same directory. Then use the provided vc10 solution file. 


History.txt
KalmanFilter application for RTPC12, By Jixie Zhang
TODO:
1) Change Bz field of each measurement layer according to field map
2) Add Bx By into the framework
3) Use Bz to help to determine head or tail for a curve back track. Currently
assumes that the field is pointing to upstream
4) Try to apply Kalman Filter twice, the first iteration is fitting from
outside to inside then smooth out, to achieve the state vector for the site at the 
end of the track. Then use this state vector as the initial value for the 2nd 
iteration, which is fitting from outside to inside. ----done by jixie, March 26 2016
5) Optimze global helix fitter. Find out in which situation it can be not used, or has
bad performance.  ---------------------------done by jixie, March 26 2016 


2017/12/06
1. Updated GEMC reader.This new reader assumes that GEMC tree uses vector to keep
   thrown parameters. The Rec tree contains hits from secondary tracks. Some
   primary tracks might not create any hits. Added variable CF_TID0[] to store the
   TrackID for valid thrown tracks. If a thrown track id not valid (valid hits < 5),
   its thrown information will not be stored. 
2. Updated Tree_fill().  Since the thrown information contains multiple track, I need
   to find the right track then store it. 

2017/09/22
1. Updated visulization for chain finder to view how a chain is found hit by hits.

2017/07/28
1. Updated visulization.
2. Solved bug in ChainFinder::SearchChains()
3. Added a measurement layer 1 mm outside the target wall to force the fit
   to include one hit close to the beam line. (not finished yet.) 
4. Added ShiftTime into the program, also store it into output tree. I plan
   to use it to indentify the real track from background tracks. Multiple files
   affected.
5. Added 'fTrueTID' into ChainFinder to identify which track is true track. Subroutine
   IdentifyThrownTID(int chainid, double &likelyhood)
   is added. I also replaced tab key by 8 spaces to reformat this class, including
   its header file.


2017/06/20
1. Updated the visulization. Added visulization to ChainFider.
2. Solved some bug about number of hits limit in ChainFinder.
3. Modified ReadGEMC since its format changed.
 

2017/06/14
1. Added visulization feature. 
   A) Added 3 base class EXVKalDetector,EXVMeasLayer and EXHYBTrack. The purpose
      is to add Draw(Int_t color, const Char_t *opt) method into Detectors and tracks.   
   B) Modified class EXKalDetector, EXMeasLayer, EXKalRTPC and EXKalManager to adapt 
      the above changes.
      
2. Modified EXKalManager to read GEMC root file as input.

3. Modified Main.cxx, use -g option to tell what type of input file. The new usage is:        
/------------------------------usage start----------------------------------------\ 
exe [-h] [-i] [-j <jobtype=0>] [-n <nevents=10000>]  [-e <error=0.05> ] \           
    [-c <ntracks=1> [space_cm=1.9] [min_ang_deg=30] [max_ang_deg=40.0] [ang_sep_cm=0.4>]] \
    [-p <pt_min_gev=0.1> <pt_max_gev=0.1>]  \                                              
    [-t <costh_min=-0.00001> <costh_max=0.00001>] \                                        
    [-z <z_min_cm=0.0> <z_max_cm=0.0>] \                                                   
    [-f <infile="infile.root">] \                                                          
    [-g <g4type=0>                                                                         

-------------------The following old usage is also supported----------------------
exe <job=0|1|2|3|4|5|6> <nevent> [pt_min_gev=0.1] [pt_max_gev=0.1] \
   [costh_min=-0.00001] [costh_max=0.00001] [error=0.05] [z_min=0.0] [z_max=0.0] \
   [infile=infile.root] [g4type=0]

          -h show this help menu;
          -i interactive mode;
          job: 0: run KalmanFilter with generated helix;
               1: run KalmanFilter with geant4 track;
               2: run KalmanFilter with generated circle;
               3: run ChainFinder but no fitting with geant4 tracks, need to couple with -c option;
               4: run ChainFinder + Global Helix Fitter with geant4 tracks
               5: run ChainFinder + KalmanFilter with geant4 tracks
          nevents: number of events to generate
          pt_min_gev and pt_max_gev: specifiy the range of pt in Gev
          Note that if pt is negative then anti-clockwise track will be generated
          costh_min and costh_max: specifiy the range of costh, only for job==0
          error is used to initialize the comvariant matrix before fitting.
          z_min and z_max are used to tell the range of thrown z.
          infile is used to specifiy the input root file.
          g4type tells the type of the input file: 0, Jixie's Geant4 RTPC12 tree, 1 GEMC tree

example1: exe 0 1000 ' -0.05' ' -0.07' ' -0.8' 0.8 0.05
example2: exe 0 1000 ' -0.05' ' -0.07' ' -0.8' 0.8 0.05 ' -20.0' 20.0 infile.root 1
example3: exe -j 0 -n 1000 -p -0.05 -0.07 -c 25 1.9 30.0 40.0 0.4 -z -20 20 -i
\------------------------------usage  end------------------------------------------/

4. Re-formated some source code.
  
   
2017/04/06
1. Added subroutine PrepareHitPool() which is passing vector as arguments.
  //Added By Carlos, using vector instead of array, provide x,x,z in cm
  void PrepareHitPool(vector<int> *id, vector<int> *tdc, vector<int> *adc, 
		      vector<double> *x, vector<double> *y,  vector<double> *z, 
		      int n, vector<int> *throwntid=0,  int append=0);
          
2017/03/15

1) Fixed one bug in EXKalRTPC::DoGlobalHelixFit(). It never do the fit before.
   Now it works.
2) Solved chi2=nan from BonusHelixFit.cc. The bug is due to the fact that include the 
   beam line point whose r=0 will cause "devide by zero".  I simple set the beam line
   point with r=0.00001 to avoid this.

3) Updated EXKalManger. Added chain-finder tree leaves for "ID, TDC, ADC, S, Phi". 
   Added subroutine  IdentifyThrownTID() to identify which thrown track a given chain 
   is corresponding to.


Main.cxx
//By Jixie Zhang: RTPC Kalman filter main program
//This program is used to test the kalman filter module
//
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <ctype.h>
#include <math.h>
#include <string.h>
#include <iostream>

#include <TROOT.h>
#include <TApplication.h>

#include "EXKalManager.h"

using namespace std;

//new usage:  exe [-h] [-i] [-j <jobtype=0>] [-n <nevents=10000>]  [-e <error=0.05> ] 
//                [-c <ntracks=1> [max_sep_cm=1.9] [max_sep_ang_deg=30] [min_sep_cm=0.4>] [min_sep_ang_deg=40.0] [ini_sep_cm=1.5>]]
//                [-p <pt_min_gev=0.1> <pt_max_gev=0.1>]  
//                [-t <costh_min=0> <costh_max=0>] 
//                [-z <z_min_cm=0> <z_max_cm=0>]   
//                [-i <infile="infile.root">]   
//also compatible to old usage:
//  exe <job=0|1|2> <nevent> [pt_min_gev=0.1] [pt_max_gev=0.1] 
//      [costh_min=-0.00001] [costh_max=0.00001] [error=0.05] [z_min=0.0] [z_max=0.0] 
//      [infile=infile.root] [g4type=0]

//void SetParameters(double max_sep, double max_sep_ang, double min_sep, double min_sep_ang,
//                   double ini_sep=1.0);
void usage()
{
  printf("/------------------------------usage start----------------------------------------\\ \n");
  printf("exe [-h] [-i] [-j <jobtype=0>] [-n <nevents=10000>]  [-e <error=0.05> ] \\\n");
  printf("    [-c <ntracks=1> [max_sep_cm=1.9] [max_sep_ang_deg=30] [min_sep_cm=0.4>] [min_sep_ang_deg=40.0] [ini_sep_cm=1.5>]] \\\n");
  printf("    [-p <pt_min_gev=0.1> <pt_max_gev=0.1>]  \\\n");
  printf("    [-t <costh_min=-0.00001> <costh_max=0.00001>] \\\n");
  printf("    [-z <z_min_cm=0.0> <z_max_cm=0.0>] \\\n");  
  printf("    [-f <infile=\"infile.root\">] \\\n");   
  printf("    [-g <g4type=0>]\n\n");  

  cerr << "-------------------The following old usage is also supported----------------------"<<endl;
  cerr <<"exe <job=0|1|2|3|4|5|6> <nevent> [pt_min_gev=0.1] [pt_max_gev=0.1] \\"<<endl 
    <<"   [costh_min=-0.00001] [costh_max=0.00001] [error=0.05] [z_min=0.0] [z_max=0.0] \\"<<endl
    <<"   [infile=infile.root] [g4type=0]"<<endl<<endl;

  cerr << "\t  -h show this help menu;\n";
  cerr << "\t  -i interactive mode;\n";
  cerr << "\t  job: 0: run KalmanFilter with generated helix;\n";
  cerr <<" \t       1: run KalmanFilter with geant4 track;\n";
  cerr << "\t       2: run KalmanFilter with generated circle;\n";
  cerr << "\t       3: run ChainFinder but no fitting with geant4 tracks, need to couple with -c option; \n";
  cerr << "\t       4: run ChainFinder + Global Helix Fitter with geant4 tracks\n";
  cerr << "\t       5: run ChainFinder + KalmanFilter with geant4 tracks\n";
  cerr << "\t  nevents: number of events to generate|read \n";
  cerr << "\t  pt_min_gev and pt_max_gev: specifiy the range of pt in Gev \n";
  cerr << "\t  Note that if pt is negative then anti-clockwise track will be generated \n";
  cerr << "\t  costh_min and costh_max: specifiy the range of costh, only for job==0\n";
  cerr << "\t  error is used to initialize the comvariant matrix before fitting. \n";
  cerr << "\t  z_min and z_max are used to tell the range of thrown z. \n";
  cerr << "\t  infile is used to specifiy the input root file. \n";
  cerr << "\t  g4type tells the type of the input file: 0, Jixie's Geant4 RTPC12 tree, 1 GEMC tree \n\n";
  cerr << "example1: exe 0 1000 ' -0.05' ' -0.07' ' -0.8' 0.8 0.05 \n";
  cerr << "example2: exe 0 1000 ' -0.05' ' -0.07' ' -0.8' 0.8 0.05 ' -20.0' 20.0 infile.root 1\n";
  cerr << "example3: exe -j 0 -n 1000 -p -0.05 -0.07 -c 25 1.9 30.0 0.4 40.0 1.5 -z -20 20 -i \n";
  cerr << "example4: exe -j 1 -n 1000 -f infile_G4MC.root -g 0 -i \n";
  cerr << "example4: exe -j 3 -n 1000 -f infile_GEMC.root -g 1 -i -c 1 1.1 30 0.4 40 0.9\n";
  cerr << "example4: exe -j 5 -n 1000 -f infile_GEMC.root -g 1 -i -c 1 \n";
  printf("\\------------------------------usage  end------------------------------------------/ \n");
  //
  printf("exe [-h] [-i] [-j <jobtype=0>] [-n <nevents=10000>]  [-e <error=0.05> ] \\\n");
  printf("    [-c <ntracks=1> [max_sep_cm=1.9] [max_sep_ang_deg=30] [min_sep_cm=0.4>] [min_sep_ang_deg=40.0] [ini_sep_cm=1.5>]] \\\n");
  printf("    [-p <pt_min_gev=0.1> <pt_max_gev=0.1>]  \\\n");
  printf("    [-t <costh_min=-0.00001> <costh_max=0.00001>] \\\n");
  printf("    [-z <z_min_cm=0.0> <z_max_cm=0.0>] \\\n");  
  printf("    [-f <infile=\"infile.root\">] \\\n");   
  printf("    [-g <g4type=0> \n\n");  
  //
  abort();
}

int main(int argc, char **argv)
{
  const double rad2deg = 180./(4.*atan(1.));

  //by default run in batch mode, will switch into interactive mode by -i option
  gROOT->SetBatch(true);

  int    job=0;
  int    nevents=10000;
  double error=0.05;
  int    ntracks=1;  //how many track to dump into hitpool in an event
  double max_sep=1.9, max_sep_ang=30.0/rad2deg, min_sep=0.4, min_sep_ang=40.0/rad2deg,ini_sep=1.5;
  double pt_min=0.1, pt_max=0.1;
  double costh_min=-0.00001, costh_max=0.00001;
  double z_min=0.0, z_max=0.0;
  char   infile[100]; sprintf(infile, "%s","infile.root");
  int    g4type=0;

  if(argc<3) usage();

  int index, c;
  int optmarker = 0;
  while ((c = getopt (argc, argv, "hij:n:e:c:p:t:z:f:x:g:")) != -1) {
    switch (c) {
    case 'h':
      usage();
      break;      
    case 'i':
      gROOT->SetBatch(false);
      break;
    case 'j':
      job = atoi(optarg);
      break;
    case 'n':
      nevents = atoi(optarg);
      break;
    case 'e':
      error = atof(optarg);
      break;
    case 'x':
      optind--;
      for(int i=0;i<5;i++){
        if ( optind < argc && (argv[optind][0]!='-' || (argv[optind][0]=='-' && isdigit(argv[optind][1]))) ) {
          //fprintf(stderr, "\n-x option: argument_%d ='%s'\n",i,argv[optind]);
          if(i==0) max_sep=atof(argv[optind]);
          if(i==1) max_sep_ang=atof(argv[optind])/rad2deg;
          if(i==2) min_sep=atof(argv[optind]);
          if(i==3) min_sep_ang=atof(argv[optind])/rad2deg;
          if(i==4) ini_sep=atof(argv[optind]);
          optind++;
        } else {  
          fprintf(stderr, "\n-x option require 5 arguments \n\n");
          usage();
        }
      }
      break;
    case 'c':
      //this is another way to do the same thing as -c
      ntracks = atoi(argv[optind-1]);
      optmarker=optind-1;
      for( ;optind < argc && (argv[optind][0]!='-' || (argv[optind][0]=='-' && isdigit(argv[optind][1]))); optind++) {
        //printf("argu[0]=%c  argu[1]=%c\n",argv[optind][0],argv[optind][1]);
        //fprintf(stderr, "\n-c option argument_%d='%s'\n",optind-optmarker,argv[optind]);
        if(optind-optmarker==1) max_sep=atof(argv[optind]);
        if(optind-optmarker==2) max_sep_ang=atof(argv[optind])/rad2deg;
        if(optind-optmarker==3) min_sep=atof(argv[optind]);
        if(optind-optmarker==4) min_sep_ang=atof(argv[optind])/rad2deg;
        if(optind-optmarker==5) ini_sep=atof(argv[optind]);
        if(optind-optmarker==5) {optind++; break;}
      }
      break;
    case 'p':
      pt_min = atof(argv[optind-1]);
      if ( optind < argc && (argv[optind][0]!='-' || (argv[optind][0]=='-' && isdigit(argv[optind][1]))) ) {
        pt_max = atof(argv[optind]);
        optind++;
      } else {
        fprintf(stderr, "\n-p option require 2 arguments \n\n");
        usage();
      }
      break;
    case 't':
      costh_min = atof(argv[optind-1]);
      if ( optind < argc && (argv[optind][0]!='-' || (argv[optind][0]=='-' && isdigit(argv[optind][1]))) ) {
        costh_max = atof(argv[optind]);
        optind++;
      } else {
        fprintf(stderr, "\n-t option require 2 arguments \n\n");
        usage();
      }
      break;
    case 'z':
      z_min = atof(argv[optind-1]);
      if ( optind < argc && (argv[optind][0]!='-' || (argv[optind][0]=='-' && isdigit(argv[optind][1]))) ) {
        z_max = atof(argv[optind]);
        optind++;
      } else {
        fprintf(stderr, "\n-z option require 2 arguments \n\n");
        usage();
      }
      break;
    case 'f':
      sprintf(infile, "%s",argv[optind-1]);
      break;
    case 'g':
      g4type = atoi(optarg);
      break;
    case '?':
      if (strchr("j:n:e:c:p:t:z:f:x:g:",optopt))
        fprintf (stderr, "Option -%c requires an argument.\n", optopt);
      else if (isprint (optopt))
        fprintf (stderr, "Unknown option `-%c'.\n", optopt);
      else
        fprintf (stderr,
        "Unknown option character `\\x%x'.\n",
        optopt);
      return -1;
    default:
      abort();
    }
  }

  //taking care of those no option arguments
  //backward compatible
  index = 0;  
  for (; optind < argc; optind++) {
    index++;
    //printf ("Non-option argument argv[%d]=%s\n", index, argv[optind]);
    if(index==1) job = atoi(argv[optind]);
    if(index==2) nevents = atoi(argv[optind]);
    if(index==3) pt_min = atof(argv[optind]);
    if(index==4) pt_max = atof(argv[optind]);
    if(index==5) costh_min = atof(argv[optind]);
    if(index==6) costh_max = atof(argv[optind]);
    if(index==7) error = atof(argv[optind]);
    if(index==8) z_min = atof(argv[optind]);
    if(index==9) z_max = atof(argv[optind]);
    if(index==10)  sprintf(infile, "%s",argv[optind]);
    if(index==11) g4type = atoi(argv[optind]);
  }

  cout<<"arguments: job="<<job<<", nevents="<<nevents<<", error="<<error<<endl 
    <<"\t pt_min="<<pt_min<<", pt_max="<<pt_max<<", costh_min="<<costh_min<<", costh_max="<<costh_max<<endl
    <<"\t z_min="<<z_min<<", z_max="<<z_max<<", infile=\'"<<infile<<"\', g4type="<<g4type<<endl
    <<"\t ntracks="<<ntracks<<", max_sep="<<max_sep<<", max_sep_ang="<<max_sep_ang*rad2deg
    <<", min_sep="<<min_sep<<", min_sep_ang="<<min_sep_ang*rad2deg<<", ini_sep="<<ini_sep<<endl
    <<endl;


  ////////////////////////////////////////////////////////////////////////////////
  EXKalManager aKalMan;
  aKalMan.SetCovMElement(error);
  aKalMan.SetG4InputFile(infile);
  int thisjob=g4type*10+job;
  if(job>=3 && job<=5) {
    aKalMan.RunCFNFit(thisjob,nevents,ntracks,max_sep,max_sep_ang,min_sep,min_sep_ang,ini_sep);
  } else{
    aKalMan.RunKF(thisjob,nevents,pt_min,pt_max,costh_min,costh_max,z_min,z_max);
  }

  return 0;
}
